modules/exploits/windows/smb/ms17_010_eternalblue.rb

```diff
@@ -42,96 +42,1012 @@ def initialize(info = {})
       [
+        Opt::RHOSTS,
         Opt::RPORT(445),
-        OptBool.new('VERIFY_TARGET', [true, 'Check if remote OS matches exploit Target.', true]),
-        OptBool.new('VERIFY_ARCH', [true, 'Check if remote architecture matches exploit Target.', true]),
         OptString.new('SMBUser', [false, '(Optional) The username to authenticate as', '']),
         OptString.new('SMBPass', [false, '(Optional) The password for the specified username', '']),
-        OptString.new('SMBDomain', [false, '(Optional) The Windows domain to use for authentication', '.'])
+        OptString.new('SMBDomain', [
+          false,
+          '(Optional) The Windows domain to use for authentication. Only affects Windows Server 2008 R2, Windows 7,' \
+            ' Windows Embedded Standard 7 target machines.',
+          ''
+        ]),
+        OptBool.new('VERIFY_TARGET', [
+          true,
+          'Check if remote OS matches exploit Target. Only affects Windows Server 2008 R2, Windows 7, Windows Embedded' \
+            ' Standard 7 target machines.',
+          true
+        ]),
+        OptBool.new('VERIFY_ARCH', [
+          true,
+          'Check if remote architecture matches exploit Target. Only affects Windows Server 2008 R2, Windows 7,' \
+            ' Windows Embedded Standard 7 target machines.',
+          true
+        ])
       ]
     )
     register_advanced_options(
       [
         OptString.new('ProcessName', [true, 'Process to inject payload into.', 'spoolsv.exe']),
-        OptInt.new('MaxExploitAttempts', [true, 'The number of times to retry the exploit.', 3]),
         OptInt.new('GroomAllocations', [true, 'Initial number of times to groom the kernel pool.', 12]),
-        OptInt.new('GroomDelta', [true, 'The amount to increase the groom count by per try.', 5])
+        OptInt.new('MaxExploitAttempts', [
+          true,
+          'The number of times to retry the exploit. Only affects Windows Server 2008 R2, Windows 7, Windows Embedded' \
+            ' Standard 7 target machines.',
+          3
+        ]),
+        OptInt.new('GroomDelta', [
+          true,
+          'The amount to increase the groom count by per try. Only affects Windows Server 2008 R2, Windows 7, Windows' \
+            ' Embedded Standard 7 target machines.',
+          5
+        ])
       ]
     )
 
   end
 
-  class EternalBlueError < StandardError
-  end
-
   def exploit
     check_code = check
-    unless check_code == CheckCode::Vulnerable || datastore['ForceExploit']
-      fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+
+    if check_code.code == 'vulnerable'
+      print_good('The target is vulnerable.')
+    else
+      print_bad('The target is not vulnerable.')
+    end
+
+    if datastore['ForceExploit'] == 'true' || check_code.code == 'vulnerable'
+      print_status('Forcing Exploit') if datastore['ForceExploit'] == 'true'
+
+      if check_code.details[:os].start_with?('Windows 2012', 'Windows 8.1', 'Windows 10 Pro Build',
```

Comment:
This should probably use the information provided by the target for the automatic selection so it's not duplicated.

Also if you used `Recog::Nizer`, you should be able to eliminate the duplicate Windows 10 and Windows 7 entries.

```
>> check_code.details
=> {:os=>"Windows 7 Professional 7601 Service Pack 1", :arch=>"x64"}
>> Recog::Nizer.match('smb.native_os', check_code.details[:os])
=> {"matched"=>"Windows 7/8 (SP + Edition)", "os.certainty"=>"1.0", "os.vendor"=>"Microsoft", "os.product"=>"Windows 7", "os.edition"=>"Professional", "os.build"=>"7601", "os.version"=>"Service Pack 1", "service.protocol"=>"smb", "fingerprint_db"=>"smb.native_os"}
```

That would off load the fingerprinting of the different build strings to Recog.

Reply:
Implemented, good shout! 👍